pkg/image/importer/importer.go

```diff
@@ -0,0 +1,739 @@
+package importer
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/golang/glog"
+	gocontext "golang.org/x/net/context"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/digest"
+	"github.com/docker/distribution/manifest/schema1"
+	"github.com/docker/distribution/registry/api/errcode"
+	"github.com/docker/distribution/registry/api/v2"
+	registryclient "github.com/docker/distribution/registry/client"
+	"github.com/docker/distribution/registry/client/auth"
+	"github.com/docker/distribution/registry/client/transport"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	kapierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/fielderrors"
+	"k8s.io/kubernetes/pkg/util/sets"
+
+	"github.com/openshift/origin/pkg/dockerregistry"
+	"github.com/openshift/origin/pkg/image/api"
+	"github.com/openshift/origin/pkg/image/api/dockerpre012"
+)
+
+const (
+	// Add a dockerregistry.Client to the passed context with this key to support v1 Docker registry importing
+	ContextKeyV1RegistryClient = "v1-registry-client"
+)
+
+// Interface loads images into an image stream import request.
+type Interface interface {
+	Import(cxt *AuthContext, isi *api.ImageStreamImport) error
+}
+
+// ImageStreamImport implements an import strategy for Docker images. It keeps a cache of images
+// per distinct auth context to reduce duplicate loads
+type ImageStreamImport struct {
```

Comment:
You can't share caches across different authentication contexts - for now,
every user + namespace has to be a unique context.  Short term this is a
cache to avoid obvious stupid duplication.

On Wed, Jan 20, 2016 at 3:25 PM, David Eads <notifications@github.com>
wrote:

> In pkg/image/importer/importer.go
> <https://github.com/openshift/origin/pull/6288#discussion_r50313400>:
>
> > +	"github.com/openshift/origin/pkg/image/api/dockerpre012"
> > +)
> > +
> > +const (
> > +	// Add a dockerregistry.Client to the passed context with this key to support v1 Docker registry importing
> > +	ContextKeyV1RegistryClient = "v1-registry-client"
> > +)
> > +
> > +// Interface loads images into an image stream import request.
> > +type Interface interface {
> > +	Import(cxt *AuthContext, isi *api.ImageStreamImport) error
> > +}
> > +
> > +// ImageStreamImport implements an import strategy for Docker images. It keeps a cache of images
> > +// per distinct auth context to reduce duplicate loads
> > +type ImageStreamImport struct {
>
> This looks like its intended to be shared, but I don't see any explicit
> information about thread safety. I'm guessing not thread safe, intended to
> be re-used on a single thread? That would mean an unshared cache which
> would have some pretty severe performance consequences for any attempt at
> parallel import.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6288/files#r50313400>.
>